core/server/common/src/main/java/alluxio/master/journal/raft/RaftSnapshotManager.java

```diff
@@ -0,0 +1,189 @@
+import org.apache.ratis.statemachine.StateMachineStorage;
+import org.apache.ratis.statemachine.impl.SimpleStateMachineStorage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.InetSocketAddress;
+import java.nio.file.Files;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/**
+ * Manages a snapshot download.
+ */
+public class RaftSnapshotManager implements AutoCloseable {
+  private static final Logger LOG = LoggerFactory.getLogger(RaftSnapshotManager.class);
+
+  private final StateMachineStorage mStorage;
+  private final Map<InetSocketAddress, RaftJournalServiceClient> mClients;
+
+  @Nullable @GuardedBy("this")
+  private CompletableFuture<Long> mDownloadFuture = null;
+
+  RaftSnapshotManager(StateMachineStorage storage) {
+    mStorage = storage;
+
+    InetSocketAddress localAddress = NetworkAddressUtils.getConnectAddress(
+        NetworkAddressUtils.ServiceType.MASTER_RPC, Configuration.global());
+    mClients = ConfigurationUtils.getMasterRpcAddresses(Configuration.global()).stream()
+        .filter(address -> !address.equals(localAddress))
+        .collect(Collectors.toMap(Function.identity(), address -> {
+          MasterSelectionPolicy policy = MasterSelectionPolicy.Factory.specifiedMaster(address);
+          return new RaftJournalServiceClient(policy);
+        }));
+  }
+
+  /**
+   * @return the log index of the last successful snapshot installation, or -1 if failure
+   */
+  public long downloadSnapshotFromFollowers() {
+    if (mDownloadFuture == null) {
+      mDownloadFuture = CompletableFuture.supplyAsync(this::core).exceptionally(err -> {
+        LOG.debug("Failed to download snapshot", err);
+        return RaftLog.INVALID_LOG_INDEX;
+      });
+    } else if (mDownloadFuture.isDone()) {
+      LOG.debug("Download operation is done");
+      Long snapshotIndex = mDownloadFuture.join();
+      LOG.debug("Retrieved downloaded snapshot at index {}", snapshotIndex);
+      mDownloadFuture = null;
+      return snapshotIndex;
+    }
+    return RaftLog.INVALID_LOG_INDEX;
+  }
+
+  private long core() {
+    List<InetSocketAddress> masterRpcAddresses =
+        ConfigurationUtils.getMasterRpcAddresses(Configuration.global());
+    InetSocketAddress localAddress = NetworkAddressUtils.getConnectAddress(
+        NetworkAddressUtils.ServiceType.MASTER_RPC, Configuration.global());
+    SnapshotInfo snapshotInfo = mStorage.getLatestSnapshot();
+    if (snapshotInfo == null) {
+      LOG.debug("No local snapshot found");
+    } else {
+      LOG.debug("Local snapshot is {}", TermIndex.valueOf(snapshotInfo.getTerm(),
+          snapshotInfo.getIndex()));
+    }
+
+    List<Map.Entry<InetSocketAddress, SnapshotMetadata>> followerInfos =
```

Comment:
A problem here is that this check will fail until a follower has a snapshot ready. So this core() function will be called over and over.

This was fixed previously by having the follower not reply until having a new enough snapshot that the master could use.

Maybe for now it is sufficient to just retry here with a timeout of 10 seconds or something like that.